pkg/querier/dispatcher.go

```diff
@@ -0,0 +1,380 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/web/api/v1/api.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors.
+
+package querier
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	prototypes "github.com/gogo/protobuf/types"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/util/annotations"
+
+	"github.com/grafana/mimir/pkg/frontend/v2/frontendv2pb"
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/querier/querierpb"
+	"github.com/grafana/mimir/pkg/streamingpromql"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/limiter"
+)
+
+var evaluateQueryRequestMessageName = proto.MessageName(&querierpb.EvaluateQueryRequest{})
+
+type Dispatcher struct {
+	engine    *streamingpromql.Engine
+	queryable storage.Queryable
+	logger    log.Logger
+}
+
+func NewDispatcher(logger log.Logger, engine *streamingpromql.Engine, queryable storage.Queryable) *Dispatcher {
+	return &Dispatcher{
+		engine:    engine,
+		queryable: queryable,
+		logger:    logger,
+	}
+}
+
+// ServeHTTP responds to requests made to the evaluation HTTP endpoint.
+// This is primarily used for debugging: most requests will arrive from query-frontends via
+// the query-scheduler over gRPC and therefore be handled by HandleProtobuf.
+//
+// Why do we call this method "ServeHTTP", and the other "HandleProtobuf"?
+// We want to satisfy the http.Handler interface, which requires a method called ServeHTTP,
+// and the querier/worker.RequestHandler interface that uses the verb "handle"
+// in its method name, so we've made the querier/worker.ProtobufRequestHandler interface use "handle" as well.
+func (d *Dispatcher) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	contentType := r.Header.Get("Content-Type")
+
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+	}
+
+	if mediaType != "application/protobuf" {
+		http.Error(w, "unsupported content type", http.StatusUnsupportedMediaType)
+		return
+	}
+
+	protoType := params["proto"]
+	if protoType == "" {
+		http.Error(w, "missing proto parameter in Content-Type header", http.StatusBadRequest)
+		return
+	}
+
+	switch protoType {
+	case evaluateQueryRequestMessageName:
+		body, err := io.ReadAll(r.Body)
```

Comment:
Should we limit the request size? Otherwise rogue user can send a very big request body?

Reply:
Given this is a temporary debugging endpoint that would never be exposed outside the cluster, I'm not too worried about this.